File: gfg-weekly-coding-contest-78/Possible Path.py
from typing import List


class Solution:
    # Version 1: BFS
    # TC: O(mn), SC: O(mn)
    def possiblePath(self, n : int, m : int, grid : List[List[int]]) -> int:
        # code here
        from collections import deque
        dis = [[float('inf')]*m for _ in range(n)]
        direction = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        queue = deque()
        queue.append([0, 0])
        dis[0][0] = 0
        while queue:
            x, y = queue.popleft()
            power = grid[x][y] == 2
            for dx, dy in direction:
                nx = x + dx
                ny = y + dy
                if 0 <= nx < n and 0 <= ny < m and dis[nx][ny] > dis[x][y] + 1:
                    if grid[nx][ny] == 1:
                        if power:
                            dis[nx][ny] = dis[x][y] + 1
                            queue.append([nx, ny])
                    else:
                        dis[nx][ny] = dis[x][y] + 1
                        queue.append([nx, ny])
        return dis[-1][-1] if dis[-1][-1] != float('inf') else -1
        

File: gfg-weekly-coding-contest-78/Possible Path.cpp

class Solution {
  public:
    // Version 1: BFS
    // TC: O(mn), SC: O(mn)
    int possiblePath(int n, int m, vector<vector<int>> &grid) {
        // code here
        vector<vector<bool>> visited(n, vector<bool>(m, false));
        deque<pair<int, int>> queue;
        queue.push_back({0, 0});
        visited[0][0] = true;
        int direction[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int ans = 0;
        while (!queue.empty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                auto item = queue.front();
                queue.pop_front();
                int x = item.first;
                int y = item.second;
                if (x == n - 1 && y == m - 1) return ans;
                bool power = grid[x][y] == 2;
                for (const auto& d : direction) {
                    int nx = x + d[0];
                    int ny = y + d[1];
                    if (0 <= nx && nx < n && 0 <= ny && ny < m && !visited[nx][ny]) {
                        if (grid[nx][ny] == 1) {
                            if (power) {
                                visited[nx][ny] = true;
                                queue.push_back({nx, ny});
                            }
                        } else {
                            visited[nx][ny] = true;
                            queue.push_back({nx, ny});
                        }
                    }
                } 
            }
            ans++;
        }
        return -1;
    }
};